### PR TITLE
Add remaining collectives to dispatcher schema

### PR DESCRIPTION
**Pull request by H-Huang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#90799 Add remaining collectives to dispatcher schema**
* #88330 Allow Process Group to support multiple backends

update ProcessGroup.hpp to call dispatcher directly

### CODE REVIEW

File: torch/csrc/distributed/c10d/Ops.cpp [hunk 1]
```diff
@@ -118,35 +19,24 @@ TORCH_LIBRARY(c10d, m) {
...
-  // __torch_dispatch__.
   m.def(
       "broadcast_(Tensor[] tensors, __torch__.torch.classes.c10d.ProcessGroup process_group, int root_rank, int root_tensor, int timeout) -> (Tensor[], __torch__.torch.classes.c10d.Work)");
   m.def(
       "allreduce_(Tensor[] tensors, __torch__.torch.classes.c10d.ProcessGroup process_group, __torch__.torch.classes.c10d.ReduceOp reduce_op, int timeout) -> (Tensor[], __torch__.torch.classes.c10d.Work)");
   m.def(
-      "allreduce_coalesced_",
-      dispatch(
-          c10::DispatchKey::CompositeExplicitAutograd, allreduce_coalesced_));
+      "allreduce_coalesced_(Tensor[] tensors, __torch__.torch.classes.c10d.ProcessGroup process_group, __torch__.torch.classes.c10d.ReduceOp reduce_op, int timeout) -> __torch__.torch.classes.c10d.Work");
```
Comment (kwen2501) [inline torch/csrc/distributed/c10d/Ops.cpp:27]: nit: can we format these long lines better?

File: torch/csrc/distributed/c10d/ProcessGroup.hpp [hunk 2]
```diff
@@ -105,173 +111,308 @@ class TORCH_API ProcessGroup : public torch::CustomClassHolder {
...
+    // backends
     if (getBackendType() == BackendType::NCCL) {
       getBackend(deviceType)->endCoalescing(reqs);
     }
   }
 
-  // Consider using ops in Ops.hpp instead of the below, which route things
-  // to the dispatcher.
-  // TODO: Find a way to force the above rule programmatically.
   virtual c10::intrusive_ptr<Work> broadcast(
```
Comment (kwen2501) [inline torch/csrc/distributed/c10d/ProcessGroup.hpp:121]: Question: do we need to think about whether these collective APIs are still virtual or not?
i.e. is there any class that still inherits from `ProcessGroup`?

File: torch/csrc/distributed/c10d/ProcessGroup.hpp [hunk 3]
```diff
@@ -105,173 +111,308 @@ class TORCH_API ProcessGroup : public torch::CustomClassHolder {
...
+                    int64_t)>();
+    // It's awakward to unbox the opts here and box them again in the custom C++
+    // op. But it's also complicated to make opts as a CustomClassHolder. Leave
+    // it as it is now.
+    return std::get<1>(op.call(
+        tensors,
+        c10::intrusive_ptr<ProcessGroup>::unsafe_reclaim_from_nonowning(this),
+        opts.rootRank,
+        opts.rootTensor,
+        opts.timeout.count()));
```
Comment (kwen2501) [inline torch/csrc/distributed/c10d/ProcessGroup.hpp:142]: I notice that we still have a copy in Ops.cpp. I wonder if we could call `ops::broadcast` here?

**Review by kwen2501 — APPROVED** [review]
Thanks @H-Huang !
It looks okay to land if it is urgent. 
Otherwise, would like to see if we can de-duplicate the code in ProcessGroup.cpp wrt to Ops.cpp (see inline comments).

File: torch/csrc/distributed/c10d/ProcessGroup.hpp (hunk 3 above)
Comment (H-Huang) [reply torch/csrc/distributed/c10d/ProcessGroup.hpp:142]: I tried that but unfortunately `Ops.hpp` also includes `ProcessGroup.hpp` so there becomes a circular dependency. Probably the follow up work is to remove the definitions in Ops.cpp (after we also update reducer.cpp)

**Pull request closed without merging**